### PR TITLE
OSAC-4203: [DEV] Proto Extensions and Fulfillment-Service Status Propagation

### PR DESCRIPTION
**Pull request by adriengentil**
## [OSAC-4203](https://redhat.atlassian.net/browse/OSAC-4203): [DEV] Proto Extensions and Fulfillment-Service Status Propagation

**Jira:** https://redhat.atlassian.net/browse/OSAC-4203
**Story type:** [DEV]
**Epic:** [OSAC-4200](https://redhat.atlassian.net/browse/OSAC-4200) — Expose NIC MAC Addresses in BareMetalInstance
**Depends on:** #410 ([OSAC-4201](https://redhat.atlassian.net/browse/OSAC-4201), merged)

### Summary

Adds `BareMetalNICStatus` and `BareMetalHardware` proto messages to both the public and private `BareMetalInstanceStatus`, then wires `syncStatus` in the BMI reconciler to propagate `object.Status.Hardware.NICs` into the private proto. The public API surfaces the field automatically through the existing proto-to-proto projection layer — no server handler changes are needed.

### Changes

**Proto** (`proto/public/` and `proto/private/`):
- Add `BareMetalNICStatus` (string mac = 1) and `BareMetalHardware` (repeated nics = 1) messages to both protos
- Add `optional BareMetalHardware hardware = 5` to public `BareMetalInstanceStatus`
- Add `optional BareMetalHardware hardware = 6` to private `BareMetalInstanceStatus` (hub = 4 already taken)
- `buf generate` — regenerated Go stubs

**Reconciler** (`internal/controllers/baremetalinstance/`):
- `syncStatus`: when `object.Status.Hardware != nil`, iterates NICs and calls `SetHardware` on the private proto status; omits the field when nil (proto3 optional semantics)

### Testing

- **Unit tests (2 new specs in Describe(syncStatus)):**
  - Hardware non-nil: asserts `hardware.nics[0].mac` propagated correctly
  - Hardware nil: asserts `HasHardware()` returns false
- All suites pass (ginkgo run -r internal); 0 lint issues

### Notes

- Integration test (AC-8, TC-NFR1-03) deferred to [OSAC-4205](https://redhat.atlassian.net/browse/OSAC-4205) — requires [OSAC-4201](https://redhat.atlassian.net/browse/OSAC-4201) CRD deployed in the kind cluster

### Acceptance Criteria

- [x] `BareMetalNICStatus` and `BareMetalHardware` in both protos
- [x] `optional BareMetalHardware hardware = N` in both `BareMetalInstanceStatus`
- [x] `buf lint && buf generate` passes
- [x] `syncStatus` maps `Hardware.NICs` to proto `hardware.nics`
- [x] GET /{id} returns `hardware.nics` when CRD has it (via projection)
- [x] GET / (List) includes `hardware.nics` (via projection)
- [x] Unit tests: populate + absent cases
- [ ] Integration test — deferred to [OSAC-4205](https://redhat.atlassian.net/browse/OSAC-4205)

## Summary by CodeRabbit

* **New Features**
  * Bare metal instance status now includes physical hardware information.
  * Network interface details, including normalized MAC addresses, are synchronized and exposed in status data.
  * Hardware information is cleared when no longer reported.

* **Bug Fixes**
  * Prevented stale hardware details from remaining after inventory data is removed.

* **Tests**
  * Added coverage for hardware synchronization, absence of hardware, and clearing outdated hardware data.

### CODE REVIEW

File: fulfillment-service/internal/controllers/baremetalinstance/baremetalinstance_reconciler_function.go [hunk 1]
```diff
@@ -470,6 +470,18 @@ func (t *task) syncStatus(object *bmfov1alpha1.BareMetalInstance) {
 	}
 	t.bareMetalInstance.GetStatus().SetNetworkAttachmentStatuses(protoStatuses)
 
+	if hw := object.Status.Hardware; hw != nil {
```
[the remark this answers is not in the document]
Comment (adriengentil) [reply fulfillment-service/internal/controllers/baremetalinstance/baremetalinstance_reconciler_function.go:474]: Fixed in `c34b7bd` — added `else { t.bareMetalInstance.GetStatus().ClearHardware() }` for consistency with the `SetNetworkAttachmentStatuses` pattern. Also added a stale-data test that calls syncStatus twice: first with hardware populated (asserts HasHardware true), then with nil hardware (asserts HasHardware false).

File: fulfillment-service/internal/controllers/baremetalinstance/baremetalinstance_reconciler_function_test.go [hunk 2]
```diff
@@ -1967,6 +1967,34 @@ var _ = Describe("syncStatus", func() {
 			privatev1.BareMetalInstanceConditionType_BARE_METAL_INSTANCE_CONDITION_TYPE_PROVISIONED)
 		Expect(cond.GetStatus()).ToNot(Equal(privatev1.ConditionStatus_CONDITION_STATUS_TRUE))
 	})
+
```
[the remark this answers is not in the document]
Comment (adriengentil) [reply fulfillment-service/internal/controllers/baremetalinstance/baremetalinstance_reconciler_function_test.go:1970]: Fixed in `c34b7bd` — added `else { t.bareMetalInstance.GetStatus().ClearHardware() }` for consistency with the `SetNetworkAttachmentStatuses` pattern. Also added a stale-data test that calls syncStatus twice: first with hardware populated (asserts HasHardware true), then with nil hardware (asserts HasHardware false).

File: fulfillment-service/internal/controllers/baremetalinstance/baremetalinstance_reconciler_function.go [hunk 3]
```diff
@@ -470,6 +470,20 @@ func (t *task) syncStatus(object *bmfov1alpha1.BareMetalInstance) {
...
+		protoNICs := make([]*privatev1.BareMetalNICStatus, 0, len(hw.NICs))
+		for _, n := range hw.NICs {
+			protoNICs = append(protoNICs, privatev1.BareMetalNICStatus_builder{
+				Mac: n.MAC,
+			}.Build())
+		}
+		t.bareMetalInstance.GetStatus().SetHardware(privatev1.BareMetalHardware_builder{
+			Nics: protoNICs,
+		}.Build())
```
[the remark this answers is not in the document]
Comment (adriengentil) [reply fulfillment-service/internal/controllers/baremetalinstance/baremetalinstance_reconciler_function.go:483]: Fixed in `28828133` — added `strings.ToLower(n.MAC)` in the syncStatus NIC builder. Updated the test to use uppercase MAC input (`"AA:BB:CC:DD:EE:01"`) and assert the stored value is lowercase (`"aa:bb:cc:dd:ee:01"`). The operator already lowercases in GetHostNICs (OSAC-4201), but syncStatus should be defensively correct regardless of CRD data casing.
